walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,212 @@
+    // Edge case if request's duration is over a day -> returns no options
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    // Get all mandatory meeting attendees & optional attendees
+    Set<String> mandatoryAttendees = new HashSet<>(request.getAttendees());
+    Set<String> optionalAttendees = new HashSet<>(request.getOptionalAttendees());
+
+    // Process which events are mandatory to get required #
+    ArrayList<TimeRange> mandatoryEvents = new ArrayList<>();
+    for (Event i : events) {
+      if (mandatoryAttendees.containsAll(i.getAttendees())) {
+        mandatoryEvents.add(i.getWhen());
+      }
+    }
+
+    if (!(mandatoryAttendees.isEmpty()) && optionalAttendees.isEmpty()) {
+      return getTimeSlotWithMandatoryAttendeesOnly(
+          mandatoryEvents, request, TimeRange.START_OF_DAY, TimeRange.END_OF_DAY);
+    } else if ((mandatoryAttendees.isEmpty() && optionalAttendees.isEmpty())) {
+      return fullDayAsOpenSlot();
+    } else {
+      return getTimeSlotWithMostOptionalAttendees(
+          events, request, mandatoryAttendees, optionalAttendees, mandatoryEvents.size());
+    }
+  }
+
+  public Collection<TimeRange> getTimeSlotWithMostOptionalAttendees(
+      Collection<Event> events,
+      MeetingRequest request,
+      Set<String> mandatoryAttendees,
+      Set<String> optionalAttendees,
+      int numOfMandatoryEvents) {
+
+    // Generate all possible intervals in O(1) time
+    int totalMinutes = 60 * 24;
+    int meetingDuration = (int) request.getDuration();
+    ArrayList<TimeRange> possibleIntervals = new ArrayList<>();
+    for (int currentMinute = 0; currentMinute < totalMinutes - meetingDuration; currentMinute++) {
+      possibleIntervals.add(
+          TimeRange.fromStartEnd(currentMinute, currentMinute + meetingDuration, false));
+    }
+
+    // Maintain a frequency of timeSlot that already works for all mandatory attendee
+    // to the # of optional attendees
+    Map<TimeRange, Integer> frequency = new HashMap<>();
+
+    int startTime = 0;
+    int optimalEndTime = 0;
+    int prevOptionalAttendee = 0;
+    int prevMandatoryAttendee = 0;
+
+    // Check all possible intervals and return the optimal one with the most optional attendees
+    for (int j = 0; j < possibleIntervals.size(); j++) {
+      TimeRange curInterval = possibleIntervals.get(j);
+      int curMandatoryAttendee = 0;
+      int curOptionalAttendee = 0;
+      second:
+      for (Event i : events) {
+        // events do not overlap
+        if (!curInterval.overlaps(i.getWhen())) {
+          if (optionalAttendees.containsAll(i.getAttendees())) {
+            curOptionalAttendee++;
+          } else if (mandatoryAttendees.containsAll(i.getAttendees())) {
+            curMandatoryAttendee++;
+          }
+        } else {
+          // events overlap
+          if (mandatoryAttendees.containsAll(i.getAttendees())) {
+            break second;
```

Comment:
You don't need this label — it's only necessary if you need to break out of two loops. In this case you only need to break out of one.

Reply:
fixed